File: api/git_info.py
"""ワークスペースの git ステータス収集（git_info）パイプライン。

複数の git クエリを並列実行し、branch / upstream / ahead-behind / diff 統計などを
1つの dict に集約する。結果は短時間 TTL でキャッシュする。

低レベルの git 実行ヘルパー（run_git_raw / _parse_github_url）は git_utils に置き、
本モジュールはそれらを import する一方向依存とする（循環 import を避ける）。
"""

import logging
import re
import subprocess
from concurrent.futures import ThreadPoolExecutor
from concurrent.futures import TimeoutError as FutureTimeoutError
from pathlib import Path
from typing import Any

from .common import (
    GIT_INFO_CACHE_TTL_SEC,
    GIT_QUICK_TIMEOUT_SEC,
    TTLCache,
    count_file_lines,
    resolve_workspace_path,
)
from .git_utils import _parse_github_url, run_git_raw

logger = logging.getLogger(__name__)

_git_info_cache = TTLCache(GIT_INFO_CACHE_TTL_SEC)
_GIT_INFO_EXECUTOR = ThreadPoolExecutor(max_workers=4, thread_name_prefix="git-info")


def invalidate_git_info(workspace_name: str):
    ws_path = resolve_workspace_path(workspace_name)
    if ws_path:
        _git_info_cache.invalidate(str(ws_path))
    else:
        _git_info_cache.invalidate(workspace_name)
    # API 経由の git 操作はここを必ず通るので、ステータスストリーム購読者へ
    # FS イベントを待たずに即時 push する（git_watch → git_info の一方向依存を
    # 保つため遅延 import）。
    from .git_watch import nudge_workspace
    nudge_workspace(workspace_name)


def refresh_git_info(directory: Path, name: str) -> dict[str, Any]:
    """watchfiles 起点の更新。diff/staged/status の3本だけ再計算してキャッシュを部分更新する。

    branch / last_commit / ahead-behind などはファイル保存では変わらないため
    キャッシュを流用し、応答を高速化する。キャッシュが無い場合はフル再計算。
    """
    cache_key = str(directory)
    cached: dict[str, Any] | None = _git_info_cache.get(cache_key)
    if cached is None:
        # 初回またはキャッシュ切れ: フル再計算
        _git_info_cache.invalidate(cache_key)
        return git_info_to_status_dict(directory, name)

    def run_git(*args):
        return run_git_raw(list(args), directory)

    _DIFF_QUERIES = {
        "status": ("--no-optional-locks", "status", "--porcelain", "--untracked-files=all"),
        "diff":   ("--no-optional-locks", "diff", "--shortstat"),
        "staged": ("--no-optional-locks", "diff", "--staged", "--shortstat"),
    }
    try:
        futures = {key: _GIT_INFO_EXECUTOR.submit(run_git, *args) for key, args in _DIFF_QUERIES.items()}
        out = {key: _stdout_if_ok(f) for key, f in futures.items()}
    except (subprocess.TimeoutExpired, OSError) as e:
        logger.warning("refresh_git_info diff queries failed dir=%s: %s", directory, e)
        return git_info_to_status_dict(directory, name)

    updated = dict(cached)
    updated["name"] = name
    if out["status"] is not None:
        updated["clean"] = len(out["status"].strip()) == 0
    # clean になった場合は diff 統計をリセット
    if updated["clean"]:
        updated["insertions"] = 0
        updated["deletions"] = 0
        updated["changed_files"] = 0
    else:
        _apply_diff_stats(updated, (out["diff"] or "", out["staged"] or ""), out["status"], directory)
    _git_info_cache.set(cache_key, {k: v for k, v in updated.items() if k != "name"})
    return updated


_FUTURE_TIMEOUT_SEC = GIT_QUICK_TIMEOUT_SEC + 2


def _safe_result(future):
    try:
        return future.result(timeout=_FUTURE_TIMEOUT_SEC)
    except (FutureTimeoutError, subprocess.TimeoutExpired, OSError) as e:
        logger.debug("git future failed: %s", e)
        return None


def _stdout_if_ok(future) -> str | None:
    r = _safe_result(future)
    return r.stdout if (r and r.returncode == 0) else None


def _empty_git_info() -> dict[str, Any]:
    return {
        "is_git_repo": False,
        "branch": None,
        "upstream": None,
        "has_upstream": None,
        "has_remote_branch": None,
        "last_commit": None,
        "last_commit_message": None,
        "github_url": None,
        "clean": None,
        "ahead": 0,
        "behind": 0,
        "insertions": 0,
        "deletions": 0,
        "changed_files": 0,
    }


def _apply_branch_and_remote(info: dict, branch_out: str | None, remote_branches_out: str | None) -> None:
    if branch_out:
        info["branch"] = branch_out.strip()
    if not info["branch"]:
        return
    if remote_branches_out:
        candidates = {b.strip() for b in remote_branches_out.splitlines() if b.strip()}
        info["has_remote_branch"] = f"origin/{info['branch']}" in candidates
    else:
        info["has_remote_branch"] = False


def _apply_upstream(info: dict, upstream_out: str | None) -> None:
    if upstream_out and upstream_out.strip():
        info["upstream"] = upstream_out.strip()
        info["has_upstream"] = True
    else:
        info["has_upstream"] = False


def _apply_diff_stats(
    info: dict,
    diff_outputs: tuple[str, ...],
    status_out: str | None,
    directory: Path,
) -> None:
    for diff_stat_output in diff_outputs:
        if not diff_stat_output:
            continue
        files_match = re.search(r"(\d+) file", diff_stat_output)
        insertions_match = re.search(r"(\d+) insertion", diff_stat_output)
        deletions_match = re.search(r"(\d+) deletion", diff_stat_output)
        if files_match:
            info["changed_files"] += int(files_match.group(1))
        if insertions_match:
            info["insertions"] += int(insertions_match.group(1))
        if deletions_match:
            info["deletions"] += int(deletions_match.group(1))
    if status_out:
        for line in status_out.splitlines():
            if line.startswith("?? "):
                info["changed_files"] += 1
                info["insertions"] += count_file_lines(directory / line[3:])


def _parse_revlist_pair(out: str) -> tuple[int, int] | None:
    parts = out.strip().split()
    if len(parts) == 2:
        try:
            return int(parts[0]), int(parts[1])
        except ValueError:
            return None
    return None


def _apply_head_commit(info: dict, commit_out: str | None, message_out: str | None) -> None:
    if commit_out and (s := commit_out.strip()):
        info["last_commit"] = s
    if message_out and (s := message_out.strip()):
        info["last_commit_message"] = s


def _apply_github_url(info: dict, remote_out: str | None) -> None:
    if remote_out and (github_url := _parse_github_url(remote_out.strip())):
        info["github_url"] = github_url


def _apply_ahead_behind(info: dict, revlist_out: str | None, run_git) -> None:
    if revlist_out and info["has_upstream"]:
        pair = _parse_revlist_pair(revlist_out)
        if pair:
            info["ahead"], info["behind"] = pair
        return
    if info["branch"] and info["has_remote_branch"] is True:
        remote_diff = run_git("rev-list", "--left-right", "--count", f"HEAD...origin/{info['branch']}")
        if remote_diff.returncode == 0:
            pair = _parse_revlist_pair(remote_diff.stdout)
            if pair:
                info["ahead"], info["behind"] = pair
        return
    if info["has_remote_branch"] is False:
        unpublished = run_git("rev-list", "--count", "HEAD", "--not", "--remotes=origin")
        if unpublished.returncode == 0:
            try:
                info["ahead"] = int(unpublished.stdout.strip() or "0")
            except ValueError:
                pass


_GIT_INFO_QUERIES: dict[str, tuple[str, ...]] = {
    "branch": ("rev-parse", "--abbrev-ref", "HEAD"),
    # 未コミット（unborn HEAD）では rev-parse が失敗するため symbolic-ref で補完する。
    "symbolic_branch": ("symbolic-ref", "--short", "HEAD"),
    "commit": ("log", "-1", "--format=%cI"),
    "message": ("log", "-1", "--format=%s"),
    "remote": ("remote", "get-url", "origin"),
    "status": ("--no-optional-locks", "status", "--porcelain", "--untracked-files=all"),
    "diff": ("--no-optional-locks", "diff", "--shortstat"),
    "staged": ("--no-optional-locks", "diff", "--staged", "--shortstat"),
    "upstream": ("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{upstream}"),
    "remote_branches": ("branch", "-r", "--format=%(refname:short)"),
    "revlist": ("rev-list", "--left-right", "--count", "HEAD...@{upstream}"),
}


def _populate_git_info(info: dict, directory: Path, run_git) -> None:
    futures = {key: _GIT_INFO_EXECUTOR.submit(run_git, *args) for key, args in _GIT_INFO_QUERIES.items()}
    out = {key: _stdout_if_ok(f) for key, f in futures.items()}

    _apply_branch_and_remote(info, out["branch"] or out["symbolic_branch"], out["remote_branches"])
    _apply_head_commit(info, out["commit"], out["message"])
    _apply_upstream(info, out["upstream"])
    _apply_github_url(info, out["remote"])

    if out["status"] is not None:
        info["clean"] = len(out["status"].strip()) == 0
    if not info["clean"]:
        _apply_diff_stats(info, (out["diff"] or "", out["staged"] or ""), out["status"], directory)

    _apply_ahead_behind(info, out["revlist"], run_git)


def git_info(directory: Path) -> dict[str, Any]:
    cache_key = str(directory)
    cached: dict[str, Any] | None = _git_info_cache.get(cache_key)
    if cached is not None:
        return cached
    info: dict[str, Any] = _empty_git_info()

    def run_git(*args):
        return run_git_raw(list(args), directory)

    try:
        check = run_git("rev-parse", "--is-inside-work-tree")
        if check.returncode != 0:
            return info
        info["is_git_repo"] = True
        _populate_git_info(info, directory, run_git)
    except (subprocess.TimeoutExpired, OSError) as e:
        logger.warning("git_info failed dir=%s: %s", directory, e)
    _git_info_cache.set(cache_key, info)
    return info


def git_info_to_status_dict(directory: Path, name: str) -> dict:
    result = git_info(directory)
    result["name"] = name
    return result

File: server/src/git_info.rs
//! ワークスペースの git ステータス収集パイプライン（Python 側 `api/git_info.py` の移植）。
//!
//! 複数の git クエリを並列実行し、branch / upstream / ahead-behind / diff 統計を
//! 1つの dict に集約する。結果は 5 秒 TTL でキャッシュし、checkout 等の直後の
//! invalidate 中に走っていた古い計算がキャッシュを上書きしないよう世代カウンタで
//! ガードする（Python と同一の設計）。

use std::collections::HashMap;
use std::path::Path;
use std::sync::Mutex;
use std::time::{Duration, Instant};

use serde_json::{json, Map, Value};

use crate::git_diff::count_file_lines;
use crate::git_utils::{parse_github_url, run_git_raw, GIT_QUICK_TIMEOUT_SEC};

const GIT_INFO_CACHE_TTL_SEC: u64 = 5;

/// キャッシュエントリ: (格納時刻, git_info の結果 dict)。
type GitInfoEntry = (Instant, Map<String, Value>);

#[derive(Default)]
pub struct GitInfoCache {
    store: Mutex<HashMap<String, GitInfoEntry>>,
    generations: Mutex<HashMap<String, u64>>,
}

impl GitInfoCache {
    pub fn new() -> Self {
        Self::default()
    }

    fn get(&self, key: &str) -> Option<Map<String, Value>> {
        let store = self.store.lock().expect("git info cache poisoned");
        store.get(key).and_then(|(ts, v)| {
            (ts.elapsed() < Duration::from_secs(GIT_INFO_CACHE_TTL_SEC)).then(|| v.clone())
        })
    }

    fn set(&self, key: &str, value: Map<String, Value>) {
        let mut store = self.store.lock().expect("git info cache poisoned");
        store.insert(key.to_string(), (Instant::now(), value));
    }

    fn generation(&self, key: &str) -> u64 {
        *self
            .generations
            .lock()
            .expect("git info generations poisoned")
            .get(key)
            .unwrap_or(&0)
    }

    /// キャッシュ無効化 + 世代を進める（Python `invalidate_git_info_cache` 相当）。
    pub fn invalidate(&self, directory: &Path) {
        let key = directory.to_string_lossy().into_owned();
        self.store
            .lock()
            .expect("git info cache poisoned")
            .remove(&key);
        *self
            .generations
            .lock()
            .expect("git info generations poisoned")
            .entry(key)
            .or_insert(0) += 1;
    }
}

fn empty_git_info() -> Map<String, Value> {
    json!({
        "is_git_repo": false,
        "branch": null,
        "upstream": null,
        "has_upstream": null,
        "has_remote_branch": null,
        "last_commit_date": null,
        "last_commit_message": null,
        "github_url": null,
        "clean": null,
        "ahead": 0,
        "behind": 0,
        "insertions": 0,
        "deletions": 0,
        "changed_files": 0,
    })
    .as_object()
    .expect("object literal")
    .clone()
}

async fn query(directory: &Path, args: &[&str]) -> Option<String> {
    match run_git_raw(args, directory, GIT_QUICK_TIMEOUT_SEC, &[]).await {
        Ok(out) if out.code == 0 => Some(out.stdout),
        _ => None,
    }
}

/// "N insertion" / "N deletion" を diff --shortstat 出力から拾う。
fn extract_stat(text: &str, keyword: &str) -> i64 {
    for (idx, _) in text.match_indices(keyword) {
        let head = &text[..idx];
        let digits: String = head
            .chars()
            .rev()
            .skip_while(|c| c.is_whitespace())
            .take_while(|c| c.is_ascii_digit())
            .collect();
        if !digits.is_empty() {
            let n: String = digits.chars().rev().collect();
            return n.parse().unwrap_or(0);
        }
    }
    0
}

fn apply_diff_stats(
    info: &mut Map<String, Value>,
    diff_outputs: &[&str],
    status_out: Option<&str>,
    directory: &Path,
) {
    let mut insertions = info["insertions"].as_i64().unwrap_or(0);
    let mut deletions = info["deletions"].as_i64().unwrap_or(0);
    let mut changed_files = info["changed_files"].as_i64().unwrap_or(0);
    for out in diff_outputs {
        if out.is_empty() {
            continue;
        }
        insertions += extract_stat(out, " insertion");
        deletions += extract_stat(out, " deletion");
    }
    if let Some(status) = status_out {
        for line in status.lines() {
            if line.is_empty() {
                continue;
            }
            changed_files += 1;
            if let Some(file) = line.strip_prefix("?? ") {
                insertions += count_file_lines(&directory.join(file));
            }
        }
    }
    info.insert("insertions".to_string(), json!(insertions));
    info.insert("deletions".to_string(), json!(deletions));
    info.insert("changed_files".to_string(), json!(changed_files));
}

fn parse_revlist_pair(out: &str) -> Option<(i64, i64)> {
    let parts: Vec<&str> = out.split_whitespace().collect();
    if parts.len() == 2 {
        Some((parts[0].parse().ok()?, parts[1].parse().ok()?))
    } else {
        None
    }
}

async fn apply_ahead_behind(
    info: &mut Map<String, Value>,
    revlist_out: Option<&str>,
    directory: &Path,
) {
    let has_upstream = info["has_upstream"] == json!(true);
    if let Some(out) = revlist_out {
        if has_upstream {
            if let Some((a, b)) = parse_revlist_pair(out) {
                info.insert("ahead".to_string(), json!(a));
                info.insert("behind".to_string(), json!(b));
            }
            return;
        }
    }
    let branch = info["branch"].as_str().unwrap_or("").to_string();
    if !branch.is_empty() && info["has_remote_branch"] == json!(true) {
        let range = format!("HEAD...origin/{branch}");
        if let Some(out) = query(directory, &["rev-list", "--left-right", "--count", &range]).await
        {
            if let Some((a, b)) = parse_revlist_pair(&out) {
                info.insert("ahead".to_string(), json!(a));
                info.insert("behind".to_string(), json!(b));
            }
        }
        return;
    }
    if info["has_remote_branch"] == json!(false) {
        if let Some(out) = query(
            directory,
            &["rev-list", "--count", "HEAD", "--not", "--remotes=origin"],
        )
        .await
        {
            if let Ok(n) = out.trim().parse::<i64>() {
                info.insert("ahead".to_string(), json!(n));
            }
        }
    }
}

async fn populate_git_info(info: &mut Map<String, Value>, directory: &Path) {
    // Python の _GIT_INFO_QUERIES と同一の 11 クエリを並列実行する
    let (
        branch,
        symbolic_branch,
        commit_date,
        message,
        remote,
        status,
        diff,
        staged,
        upstream,
        remote_branches,
        revlist,
    ) = tokio::join!(
        query(directory, &["rev-parse", "--abbrev-ref", "HEAD"]),
        query(directory, &["symbolic-ref", "--short", "HEAD"]),
        query(directory, &["log", "-1", "--format=%cI"]),
        query(directory, &["log", "-1", "--format=%s"]),
        query(directory, &["remote", "get-url", "origin"]),
        query(
            directory,
            &[
                "--no-optional-locks",
                "status",
                "--porcelain",
                "--untracked-files=all"
            ]
        ),
        query(directory, &["--no-optional-locks", "diff", "--shortstat"]),
        query(
            directory,
            &["--no-optional-locks", "diff", "--staged", "--shortstat"]
        ),
        query(
            directory,
            &[
                "rev-parse",
                "--abbrev-ref",
                "--symbolic-full-name",
                "@{upstream}"
            ]
        ),
        query(directory, &["branch", "-r", "--format=%(refname:short)"]),
        query(
            directory,
            &["rev-list", "--left-right", "--count", "HEAD...@{upstream}"]
        ),
    );

    // branch（unborn HEAD は symbolic-ref で補完）+ リモートブランチ有無
    let branch_out = branch.or(symbolic_branch);
    if let Some(b) = branch_out
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty())
    {
        info.insert("branch".to_string(), json!(b));
    }
    if info["branch"].is_string() {
        let branch_name = info["branch"].as_str().unwrap_or("").to_string();
        let has_remote = remote_branches
            .as_deref()
            .map(|out| {
                let target = format!("origin/{branch_name}");
                out.lines().map(str::trim).any(|c| c == target)
            })
            .unwrap_or(false);
        info.insert("has_remote_branch".to_string(), json!(has_remote));
    }

    if let Some(s) = commit_date
        .as_deref()
        .map(str::trim)
        .filter(|s| !s.is_empty())
    {
        info.insert("last_commit_date".to_string(), json!(s));
    }
    if let Some(s) = message.as_deref().map(str::trim).filter(|s| !s.is_empty()) {
        info.insert("last_commit_message".to_string(), json!(s));
    }

    match upstream.as_deref().map(str::trim).filter(|s| !s.is_empty()) {
        Some(u) => {
            info.insert("upstream".to_string(), json!(u));
            info.insert("has_upstream".to_string(), json!(true));
        }
        None => {
            info.insert("has_upstream".to_string(), json!(false));
        }
    }

    if let Some(url) = remote.as_deref().and_then(|r| parse_github_url(r.trim())) {
        info.insert("github_url".to_string(), json!(url));
    }

    if let Some(s) = status.as_deref() {
        info.insert("clean".to_string(), json!(s.trim().is_empty()));
    }
    if info["clean"] != json!(true) {
        apply_diff_stats(
            info,
            &[
                diff.as_deref().unwrap_or(""),
                staged.as_deref().unwrap_or(""),
            ],
            status.as_deref(),
            directory,
        );
    }

    apply_ahead_behind(info, revlist.as_deref(), directory).await;
}

/// git ステータス一式（Python `git_info` 相当 — TTL キャッシュ + 世代ガード付き）。
pub async fn git_info(cache: &GitInfoCache, directory: &Path) -> Map<String, Value> {
    let cache_key = directory.to_string_lossy().into_owned();
    if let Some(cached) = cache.get(&cache_key) {
        return cached;
    }
    let gen_at_start = cache.generation(&cache_key);
    let mut info = empty_git_info();
    let is_repo = query(directory, &["rev-parse", "--is-inside-work-tree"])
        .await
        .is_some();
    if is_repo {
        info.insert("is_git_repo".to_string(), json!(true));
        populate_git_info(&mut info, directory).await;
    }
    // 計算中に invalidate（世代が進んだ）されていたら古い結果でキャッシュを
    // 上書きしない（呼び出し元にはそのまま返す）
    if cache.generation(&cache_key) == gen_at_start {
        cache.set(&cache_key, info.clone());
    }
    info
}

/// /workspaces/statuses の要素形式（`git_info_to_status_json` 相当）。
pub async fn git_info_to_status_json(cache: &GitInfoCache, directory: &Path, name: &str) -> Value {
    let mut result = git_info(cache, directory).await;
    result.insert("name".to_string(), json!(name));
    Value::Object(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shortstat_extraction() {
        let out = " 2 files changed, 10 insertions(+), 3 deletions(-)\n";
        assert_eq!(extract_stat(out, " insertion"), 10);
        assert_eq!(extract_stat(out, " deletion"), 3);
        assert_eq!(
            extract_stat(" 1 file changed, 1 insertion(+)\n", " insertion"),
            1
        );
        assert_eq!(
            extract_stat(" 1 file changed, 1 insertion(+)\n", " deletion"),
            0
        );
        assert_eq!(extract_stat("", " insertion"), 0);
    }

    #[test]
    fn revlist_pair_parse() {
        assert_eq!(parse_revlist_pair("2\t3\n"), Some((2, 3)));
        assert_eq!(parse_revlist_pair("garbage"), None);
    }

    fn sh_git(repo: &Path, args: &[&str]) {
        let out = std::process::Command::new("git")
            .args(args)
            .current_dir(repo)
            .output()
            .unwrap();
        assert!(
            out.status.success(),
            "git {args:?}: {}",
            String::from_utf8_lossy(&out.stderr)
        );
    }

    #[tokio::test]
    async fn git_info_on_real_repo() {
        let dir = tempfile::tempdir().unwrap();
        let repo = dir.path();
        sh_git(repo, &["init", "-q", "-b", "main"]);
        sh_git(repo, &["config", "user.email", "t@example.com"]);
        sh_git(repo, &["config", "user.name", "t"]);
        std::fs::write(repo.join("a.txt"), "x\n").unwrap();
        sh_git(repo, &["add", "-A"]);
        sh_git(repo, &["commit", "-q", "-m", "first"]);
        std::fs::write(repo.join("a.txt"), "x\ny\n").unwrap();
        std::fs::write(repo.join("new.txt"), "1\n2\n").unwrap();

        let cache = GitInfoCache::new();
        let info = git_info(&cache, repo).await;
        assert_eq!(info["is_git_repo"], true);
        assert_eq!(info["branch"], "main");
        assert_eq!(info["clean"], false);
        assert_eq!(info["last_commit_message"], "first");
        assert_eq!(info["changed_files"], 2);
        // 変更 1 行 + 未追跡 2 行
        assert_eq!(info["insertions"], 3);
        assert_eq!(info["has_upstream"], false);
        assert_eq!(info["has_remote_branch"], false);

        // キャッシュが効く（同一 Map が返る）
        let again = git_info(&cache, repo).await;
        assert_eq!(again, info);

        // invalidate → クリーンにした状態が反映される
        sh_git(repo, &["add", "-A"]);
        sh_git(repo, &["commit", "-q", "-m", "second"]);
        cache.invalidate(repo);
        let after = git_info(&cache, repo).await;
        assert_eq!(after["clean"], true);
        assert_eq!(after["last_commit_message"], "second");

        // 非 git ディレクトリは is_git_repo=false
        let plain = tempfile::tempdir().unwrap();
        let info = git_info(&cache, plain.path()).await;
        assert_eq!(info["is_git_repo"], false);
        assert_eq!(info["branch"], Value::Null);
    }
}
